File: tracker.js
// For every folder in the writing folder, compare the word count with the
// file in the same name in the snapshot folder. If there is no copy in the
// snapshot folder, pretend its word count is 0.

var fs = require('fs');
var glob = require('glob');
var wc = require('word-count');
var path = require('path');
var moment = require('moment');

var writing_folder = "/home/leonardo/Dropbox/Notes/Escrita/";
var snapshot_folder = writing_folder + "snapshot/";
var logFile = "/home/leonardo/Dropbox/Notes/quantifiedself/writing.csv";

var countCurrentFile = function(file) {
  var content = fs.readFileSync(file, 'utf8');
  var count;

  if (content == null || content.length == 0) count = 0;
  else count = wc(content);

  return count;
}

var currentFilePath = function(file) {
  return path.join(writing_folder, file);
}

var oldFilePath = function(file) {
  return path.join(snapshot_folder, path.basename(file));
}

var countOldFile = function(file) {
  var path = oldFilePath(file);

  if (fs.existsSync(path)) {
    var content = fs.readFileSync(oldFilePath(file), 'utf8');
    if (content == null || content.length == 0) count = 0;
    else count = wc(content);
  } else {
    count = 0;
  }

  return count;
}

var wordCountDifferenceForFile = function(file) {
  return countCurrentFile(file) - countOldFile(file);
}

var clearSnapshot = function() {
  glob(path.join(snapshot_folder, "*.{txt,md}"), null, function(err, files) {
    files.forEach(function(file) {
      console.log('removendo ' + file);
      fs.unlinkSync(file);
    });
  });
};

var copyFileToSnapshot = function(file) {
  fs.createReadStream(file).pipe(fs.createWriteStream(oldFilePath(file)));
};

var logWordCount = function(wordCount) {
  var entry = dateTime() + "," + wordCount + "\n";
  fs.appendFile(logFile, entry, function(err) {
   if (err) throw err; 
  });
};

var dateTime = function() {
  return moment().format('YYYY-MM-DD HH:mm:ss');
};




var wordCount = 0;
glob(path.join(writing_folder, "*.{txt,md}"), null, function(err, files) {
  // First, count words by comparing to the snapshot
  files.forEach(function(file) {
    wordCount += wordCountDifferenceForFile(file);
  });

  // Then delete everything in the snapshot
  clearSnapshot();

  // And then we copy all those files to the snapshot for tomorrow's comparison
  files.forEach(function(file) {
    copyFileToSnapshot(file);
  });

  // We finish by writing today's writing count to our CSV log file
  logWordCount(wordCount);
});

File: tracker.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'bundler'
Bundler.require(:default)

@writing_folder = "/home/leonardo/Dropbox/Notes/Escrita"
@snapshot_folder = @writing_folder + "/snapshot"
@log_file = "/home/leonardo/Dropbox/Notes/quantifiedself/writing.csv"

def old_file_path(file)
  inode = File.stat(file).ino
  old_file_name = inode.to_s + ".txt"
  File.join(@snapshot_folder, old_file_name)
end

def count_old_file_words(file)
  begin
    count = WordsCounted.from_file(old_file_path(file)).token_count
  rescue
    count = 0
  end

  count
end

def count_current_file_words(file)
  WordsCounted.from_file(file).token_count
end

def count_word_difference(file)
  count_current_file_words(file) - count_old_file_words(file)
end

def clear_snapshot_folder
  Dir.glob(File.join(@snapshot_folder, "*.{txt,md}")).each do |file|
    File.delete(file)
  end
end

def log_word_count

end

def copy_file_to_snapshot_folder(file)
  FileUtils.cp(file, old_file_path(file))
end

def get_current_files
  files = []
  Dir.glob(@writing_folder + "/*.{txt,md,markdown}").each do |f|
    files << f
  end

  files
end

def count_todays_words
  word_count = 0
  files = get_current_files

  files.each do |file|
    word_count = word_count + count_word_difference(file)
  end

  word_count
end


puts count_todays_words
clear_snapshot_folder

get_current_files.each do |file|
  copy_file_to_snapshot_folder(file)
end

